### PR TITLE
Continued: deepxde.data.pde.PDE: pde function third parameter type changed from numpy.array to tensor

### PR DESCRIPTION
**Pull request by smao-astro**
See #179 for detailed discussion.

## Improvement

- Change `coefficients_generator` to `auxiliary_var_function`, `coefficients` to `auxiliary_vars`.
- Move most of the modification of `deepxde/maps/fnn.py` to `deepxde/maps/map.py`.
- Merge `CoefficientsPDE` into `deepxde.data.pde.PDE`.
- Add support of `auxiliary_var_function` to `deepxde.data.pde.TimePDE`.

## Test

tested use the script below, `examples/Burgers.py`, `examples/diffusion_1d_resample.py`
```python
#!/usr/bin/env python
# coding: utf-8

# <a href="https://colab.research.google.com/github/lululxvi/deepxde/blob/master/examples/Lorenz_inverse_forced_Colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# # Description
# 
# This notebook aims at the identification of the parameters of the modified Lorenz attractor (with exogenous input)
# 
# Built upon: 
# * Lorenz attractor example from DeepXDE (Lu's code)
# * https://github.com/lululxvi/deepxde/issues/79
# * kind help from Lu, greatly acknowledged
# 
# # Install lib and imports

# In[1]:


# !pip install deepxde

from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import sys
import re
import numpy as np
import requests
import io
import matplotlib.pyplot as plt

sys.path.insert(0, "/Users/kyika/project/pinn/deepxde")

import deepxde as dde
from deepxde.backend import tf

import scipy as sp
import scipy.interpolate as interp
from scipy.integrate import odeint


# # Generate data

# In[2]:


# true values, see p. 15 in https://arxiv.org/abs/1907.04502
C1true = 10
C2true = 15
C3true = 8 / 3

# time points
maxtime = 3
time = np.linspace(0, maxtime, 200)
ex_input = 10 * np.sin(2 * np.pi * time)  # exogenous input

# interpolate time / lift vectors (for using exogenous variable without fixed time stamps)
def ex_func(t):
    spline = sp.interpolate.Rbf(
        time, ex_input, function="thin_plate", smooth=0, episilon=0
    )
    # return spline(t[:,0:])
    return spline(t)


# function that returns dy/dt
def LorezODE(x, t):  # Modified Lorenz system (with exogenous input).
    x1, x2, x3 = x
    dxdt = [
        C1true * (x2 - x1),
        x1 * (C2true - x3) - x2,
        x1 * x2 - C3true * x3 + ex_func(t),
    ]
    return dxdt


# initial condition
x0 = [-8, 7, 27]

# solve ODE
x = odeint(LorezODE, x0, time)

# plot results
plt.figure()
plt.plot(time, x, time, ex_input)
plt.xlabel("time")
plt.ylabel("x(t)")
# plt.show()
plt.savefig("Lorenz_inverse_forced_Colab_fig1.png", format="png")

time = time.reshape(-1, 1)
time.shape


# # Perform identification

# In[3]:


# parameters to be identified
C1 = tf.Variable(1.0)
C2 = tf.Variable(1.0)
C3 = tf.Variable(1.0)

# interpolate time / lift vectors (for using exogenous variable without fixed time stamps)
def ex_func2(x):
    spline = sp.interpolate.Rbf(
        time, ex_input, function="thin_plate", smooth=0, episilon=0
    )
    return spline(x)
    # return spline(t)


# define system ODEs
def Lorenz_system(x, y, c):
    """Modified Lorenz system (with exogenous input).
    dy1/dx = 10 * (y2 - y1)
    dy2/dx = y1 * (28 - y3) - y2
    dy3/dx = y1 * y2 - 8/3 * y3 + u
    """
    y1, y2, y3 = y[:, 0:1], y[:, 1:2], y[:, 2:]
    dy1_x = dde.grad.jacobian(y, x, i=0)
    dy2_x = dde.grad.jacobian(y, x, i=1)
    dy3_x = dde.grad.jacobian(y, x, i=2)
    return [
        dy1_x - C1 * (y2 - y1),
        dy2_x - y1 * (C2 - y3) + y2,
        dy3_x - y1 * y2 + C3 * y3 - c,
        # dy3_x - y1 * y2 + C3 * y3 - 10*tf.math.sin(2*np.pi*x),
    ]


def boundary(_, on_initial):
    return on_initial


# define time domain
geom = dde.geometry.TimeDomain(0, maxtime)

# Initial conditions
ic1 = dde.IC(geom, lambda X: x0[0], boundary, component=0)
ic2 = dde.IC(geom, lambda X: x0[1], boundary, component=1)
ic3 = dde.IC(geom, lambda X: x0[2], boundary, component=2)

# Get the training data
observe_t, ob_y = time, x

ptset = dde.bc.PointSet(observe_t)
inside = lambda x, _: ptset.inside(x)

# boundary conditions
observe_y0 = dde.DirichletBC(
    geom, ptset.values_to_func(ob_y[:, 0:1]), inside, component=0
)
observe_y1 = dde.DirichletBC(
    geom, ptset.values_to_func(ob_y[:, 1:2]), inside, component=1
)
observe_y2 = dde.DirichletBC(
    geom, ptset.values_to_func(ob_y[:, 2:3]), inside, component=2
)

# define data object
data = dde.data.PDE(
    geom,
    Lorenz_system,
    [ic1, ic2, ic3, observe_y0, observe_y1, observe_y2],
    num_domain=400,
    num_boundary=2,
    anchors=observe_t,
    auxiliary_var_function=ex_func2
)

plt.figure()
plt.plot(observe_t, ob_y)
plt.xlabel("Time")
plt.legend(["x", "y", "z"])
plt.title("Training data")
# plt.show()
plt.savefig("Lorenz_inverse_forced_Colab_fig2.png", format="png")

# define FNN architecture and compile
net = dde.maps.FNN([1] + [40] * 3 + [3], "tanh", "Glorot uniform")
model = dde.Model(data, net)
model.compile("adam", lr=0.001)

# callbacks for storing results
fnamevar = "variables.dat"
variable = dde.callbacks.VariableValue([C1, C2, C3], period=1, filename=fnamevar)

losshistory, train_state = model.train(epochs=60000, callbacks=[variable])


# Plots

# In[ ]:


# reopen saved data using callbacks in fnamevar
lines = open(fnamevar, "r").readlines()

# read output data in fnamevar (this line is a long story...)
Chat = np.array(
    [
        np.fromstring(
            min(re.findall(re.escape("[") + "(.*?)" + re.escape("]"), line), key=len),
            sep=",",
        )
        for line in lines
    ]
)

l, c = Chat.shape

plt.figure()
plt.plot(range(l), Chat[:, 0], "r-")
plt.plot(range(l), Chat[:, 1], "k-")
plt.plot(range(l), Chat[:, 2], "g-")
plt.plot(range(l), np.ones(Chat[:, 0].shape) * C1true, "r--")
plt.plot(range(l), np.ones(Chat[:, 1].shape) * C2true, "k--")
plt.plot(range(l), np.ones(Chat[:, 2].shape) * C3true, "g--")
plt.legend(["C1hat", "C2hat", "C3hat", "True C1", "True C2", "True C3"], loc="right")
plt.xlabel("Epoch")
# plt.show()
plt.savefig("Lorenz_inverse_forced_Colab_fig3.png", format="png")


yhat = model.predict(observe_t)

plt.figure()
plt.plot(observe_t, ob_y, "-", observe_t, yhat, "--")
plt.xlabel("Time")
plt.legend(["x", "y", "z", "xh", "yh", "zh"])
plt.title("Training data")
# plt.show()
plt.savefig("Lorenz_inverse_forced_Colab_fig4.png", format="png")

```

### CODE REVIEW

**Comment by smao-astro** [comment]
Hi Lu,

I am wondering whether or not users are allowed and recommended to do this
```python
# define system ODEs
def Lorenz_system(x, y,):
    """Modified Lorenz system (with exogenous input).
    dy1/dx = 10 * (y2 - y1)
    dy2/dx = y1 * (28 - y3) - y2
    dy3/dx = y1 * y2 - 8/3 * y3 + u
    """
    y1, y2, y3 = y[:, 0:1], y[:, 1:2], y[:, 2:]
    dy1_x = dde.grad.jacobian(y, x, i=0)
    dy2_x = dde.grad.jacobian(y, x, i=1)
    dy3_x = dde.grad.jacobian(y, x, i=2)
    return [
        dy1_x - C1 * (y2 - y1),
        dy2_x - y1 * (C2 - y3) + y2,
        dy3_x - y1 * y2 + C3 * y3 - tf.numpy_function(ex_func2, [tf.stop_gradient(x)], tf.float32),
        # dy3_x - y1 * y2 + C3 * y3 - 10*tf.math.sin(2*np.pi*x),
    ]
```
to directly use `net.inputs`, a tensor, as the input to a numpy function, and make this a operator in the graph, this seems attempting to me, but I can not get it run on my laptop and getting

> Initializing variables...
> Training model...
> 
> Step      Train loss                                                                                    Test loss                                                                                     Test metric
> 0         [7.65e-02, 6.16e-02, 4.94e+01, 6.40e+01, 4.90e+01, 7.29e+02, 3.26e+01, 4.08e+01, 1.96e+02]    [7.65e-02, 6.16e-02, 4.94e+01, 0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00]    []  
> 
> Process finished with exit code 139 (interrupted by signal 11: SIGSEGV)

File: deepxde/data/pde.py [hunk 1]
```diff
@@ -65,10 +66,13 @@ def __init__(
         self.soln = solution
         self.num_test = num_test
 
+        self.auxiliary_var_fn = auxiliary_var_function
```
Comment (lululxvi) [inline deepxde/data/pde.py:69]: Add `auxiliary_var_function` and `self.train_aux_vars` to the docstring

File: deepxde/data/pde.py [hunk 2]
```diff
@@ -121,41 +121,41 @@ def losses_train():
...
         self.test_y = self.soln(self.test_x) if self.soln else None
-        return self.test_x, self.test_y
+        if self.auxiliary_var_fn is not None:
+            self.test_aux_vars = self.auxiliary_var_fn(self.test_x)
+        return self.test_x, self.test_y, self.test_aux_vars
 
     def resample_train_points(self):
         """Resample the training residual points.
 
         Warning: After resampling, need to call ``Model.compile()`` to update the loss.
```
Comment (lululxvi) [inline deepxde/data/pde.py:156]: Remove the warning in docstring

File: deepxde/maps/map.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 
 from ..backend import tf
 from ..utils import make_dict, timing
+from .. import config
```
Comment (lululxvi) [inline deepxde/maps/map.py:7]:
```python
from .. import config
from ..backend import tf
from ..utils import make_dict, timing
```

File: deepxde/maps/map.py [hunk 4]
```diff
@@ -13,6 +14,7 @@ def __init__(self):
         self.training = tf.placeholder(tf.bool)
         self.dropout = tf.placeholder(tf.bool)
         self.data_id = tf.placeholder(tf.uint8)  # 0: train data, 1: test data
+        self._auxiliary_vars = None
```
Comment (lululxvi) [inline deepxde/maps/map.py:17]:
```python
self._auxiliary_vars = tf.placeholder(config.real(tf))
```

Then remove it from `build()`, and no need to call `super(FNN, self).build()` etc

File: deepxde/maps/map.py [hunk 5]
```diff
@@ -42,7 +44,14 @@ def outputs(self):
     def targets(self):
         """Targets of the mapping outputs."""
 
-    def feed_dict(self, training, dropout, data_id, inputs, targets=None):
+    @property
+    def auxiliary_vars(self):
+        """Any additional coefficients needed."""
```
Comment (lululxvi) [inline deepxde/maps/map.py:49]: """Any additional variables needed."""

File: deepxde/maps/map.py [hunk 6]
```diff
@@ -75,4 +89,5 @@ def apply_output_transform(self, transform):
     @timing
     def build(self):
         """Construct the mapping."""
+        self._auxiliary_vars = tf.placeholder(config.real(tf))
```
Comment (lululxvi) [inline deepxde/maps/map.py:92]: Delete

File: deepxde/maps/mfnn.py [hunk 7]
```diff
@@ -51,6 +51,7 @@ def targets(self):
     @timing
     def build(self):
         print("Building multifidelity neural network...")
+        super(MfNN, self).build()
```
Comment (lululxvi) [inline deepxde/maps/mfnn.py:54]: Delete

File: deepxde/maps/opnn.py [hunk 8]
```diff
@@ -94,6 +94,7 @@ def _feed_dict_inputs(self, inputs):
     @timing
     def build(self):
         print("Building operator neural network...")
+        super(OpNN, self).build()
```
Comment (lululxvi) [inline deepxde/maps/opnn.py:97]: Delete

File: deepxde/maps/resnet.py [hunk 9]
```diff
@@ -49,6 +49,7 @@ def targets(self):
     @timing
     def build(self):
         print("Building residual neural network...")
+        super(ResNet, self).build()
```
Comment (lululxvi) [inline deepxde/maps/resnet.py:52]: Delete

**Comment by lululxvi** [comment]
`tf.numpy_function` seems a good idea, but I have no experience with this function. I Will test it later.

File: deepxde/data/pde.py (hunk 2 above)
Comment (smao-astro) [reply deepxde/data/pde.py:156]: I guess we still need to call ``Model.compile()`` because numpy array is still being used in `bc.error`.
https://github.com/lululxvi/deepxde/blob/2f4a174962e4162881e2a4737deb96c92b01b980/deepxde/data/pde.py#L114

File: deepxde/data/pde.py [hunk 10]
```diff
@@ -25,6 +25,7 @@ class PDE(Data):
             points.
         solution: The reference solution.
         num_test: The number of residual points for testing the PDE residual.
+        auxiliary_var_function: A function that inputs `train_x` or `test_x` and output auxiliary variables.
```
Comment (lululxvi) [inline deepxde/data/pde.py:28]: output --> outputs
